### PR TITLE
remote-week2:user api

### PR DESCRIPTION
**Pull request by mcoba891120**
Elastic IP : 13.112.217.190
Database hostname : mcoba-db.cyldhxt4vckd.ap-northeast-1.rds.amazonaws.com

Change Description:
two apps added:
1. app.js for healthchek at port 3000
2. userapi.js at port 3001 
image
image

- Synchronous functions execute one task at a time, blocking the execution of the next task until the current task is complete like the image below. Asynchronous functions, on the other hand, allow the program to continue executing while the function is running. From above we can know that Asynchronous function is more efficient in time.
- Asynchronous functions are most useful in situations where the program needs to perform time-consuming tasks without blocking the main execution thread.
- Callbacks are simple but can be difficult to read and maintain, like having a callback hell. Promises provide a more readable way to handle asynchronous operations but require additional code to handle errors.
Async/await is the most readable and maintainable approach but requires a solid understanding of Promises to use effectively or might result in blocking. Also, Async/await is not supported by older browsers and Node.js versions.

### CODE REVIEW

File: userapi/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (addie-tyc) [inline userapi/.gitignore]: .gitignore 通常只會在一個 repo 加一個呦，package.json 也是

File: userapi/package.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "dependencies": {
+    "body-parser": "^1.20.2",
```
Comment (addie-tyc) [inline userapi/package.json:3]: 你的專案中安裝了 bodyParser，但我看你用的這個 express 版本應該有內建 urlencoded，可以確認一下這件事嗎？

File: userapi/userapi.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+app.get('/', (req, res) => {
+    res.send('Hello, World!');
+});
+
+app.listen(3001, () => {
+    console.log('Server is running on port 3001');
+
+});
+app.get('/users/:id', (req, res) => {
+    const userId = req.params.id
```
Comment (addie-tyc) [inline userapi/userapi.js]: API 文件規定是 query parameter 喔

File: userapi/userapi.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+});
+connection.query('SELECT * FROM users',(error, results, fields) => {
+    if (error) {
+        console.error('Error executing query: ' + error.stack);
+        return;
+    }
+
+    console.log('Query results: ', results)
+});
```
Comment (addie-tyc) [inline userapi/userapi.js]: 這些小測試建議是單獨寫，不然每次你啟動 app 都會多跑這一段

File: userapi/userapi.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        useremail = rows.email
+        date = rows.created
+        res.json({
+            data:{
+                user:{
+                    id:userId,
+                    name:username,
+                    email:useremail,
+                },
+                date:date
```
Comment (addie-tyc) [inline userapi/userapi.js]: date 要從 headers 拿呦，詳見 API 文件

File: userapi/userapi.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            }
+        });
+    });
+});
+
+app.use(bodyParser.json());
+app.post('/users',(req, res) => {
+    const {name, email, password} = req.body;
+    const datedate  = new Date();
+    const date = datedate.toUTCString()
```
Comment (addie-tyc) [inline userapi/userapi.js]: date 要從 headers 拿呦，詳見 API 文件

File: myapp/package.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "express": "^4.18.2",
+    "serverless-http": "^3.1.1"
```
Comment (addie-tyc) [inline myapp/package.json:13]: 這個 module 有用到嗎？

**Review by addie-tyc — CHANGES_REQUESTED** [review]
問題修正後再通知我一次呦～

File: userapi/userapi.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            res.status(400).json({error: 'database error'});
+            return;
+        }
+        if (rows.length === 0) {
+            res.status(403).json({error: 'user not found'});
+            return;
+        }
+        username = rows.name
+        useremail = rows.email
+        date = rows.created
```
Comment (addie-tyc) [inline userapi/userapi.js]: 沒改到呦

File: userapi/userapi.js [hunk 9]
```diff
@@ -78,8 +79,7 @@ app.get('/users/:id', (req, res) => {
 app.use(bodyParser.json());
 app.post('/users',(req, res) => {
     const {name, email, password} = req.body;
-    const datedate  = new Date();
-    const date = datedate.toUTCString()
+    const date = new Date().toUTCString();
```
Comment (addie-tyc) [inline userapi/userapi.js]: 這裡也沒改到呦

**Review by addie-tyc — CHANGES_REQUESTED** [review]
headers~~

File: userapi/userapi.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        if (error) {
+            res.status(400).json({error:'internal server error'});
+            return;
+        }
+        const count = results[0].count;
+        if (count > 0){
+            res.status(403).json({error:'email already exists'});
+            return;
+        }
+        connection.query('INSERT INTO users (name, email, password, created) VALUES (?,?,?,?)', [name, email, password,date_created],(error,results,fields) => {
```
Comment (addie-tyc) [inline userapi/userapi.js]: 沒有檢查 email 的 format

Comment (addie-tyc) [reply userapi/userapi.js]: 名字也是

**Review by addie-tyc — APPROVED** [review]
確定檢查到，測試過後（把測試的照片貼上來），就 merge 吧

**Comment by mcoba891120** [comment]
截圖 2023-02-27 上午12 48 03
截圖 2023-02-27 上午12 46 59
截圖 2023-02-27 上午12 45 53
截圖 2023-02-27 上午12 45 24
截圖 2023-02-27 上午12 42 55

**Pull request merged**